File: ios/runanywhere-sdks-latest/examples/flutter/RunAnywhereAI/lib/features/models/model_list_view_model.dart
import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:runanywhere/runanywhere.dart' as sdk;

import 'package:runanywhere_ai/features/models/model_types.dart';

/// ModelListViewModel (mirroring iOS ModelListViewModel.swift)
///
/// Manages model loading, selection, and state.
/// Now properly fetches models from the SDK registry and uses SDK for downloads.
class ModelListViewModel extends ChangeNotifier {
  static final ModelListViewModel shared = ModelListViewModel._();

  ModelListViewModel._() {
    unawaited(_initialize());
  }

  // State
  List<ModelInfo> _availableModels = [];
  List<LLMFramework> _availableFrameworks = [];
  ModelInfo? _currentModel;
  bool _isLoading = false;
  String? _errorMessage;

  // Download progress tracking
  final Map<String, double> _downloadProgress = {};
  final Set<String> _downloadingModels = {};

  // Getters
  List<ModelInfo> get availableModels => _availableModels;
  List<LLMFramework> get availableFrameworks => _availableFrameworks;
  ModelInfo? get currentModel => _currentModel;
  bool get isLoading => _isLoading;
  String? get errorMessage => _errorMessage;
  Map<String, double> get downloadProgress =>
      Map.unmodifiable(_downloadProgress);
  bool isDownloading(String modelId) => _downloadingModels.contains(modelId);

  Future<void> _initialize() async {
    await loadModelsFromRegistry();
  }

  /// Load models from SDK registry
  /// Fetches all registered models from the RunAnywhere SDK
  Future<void> loadModelsFromRegistry() async {
    _isLoading = true;
    _errorMessage = null;
    notifyListeners();

    try {
      // Get all models from SDK registry
      final sdkModels = await sdk.RunAnywhere.availableModels();

      // Convert SDK ModelInfo to app ModelInfo
      _availableModels = sdkModels.map(_convertSDKModel).toList();

      debugPrint(
          '✅ Loaded ${_availableModels.length} models from SDK registry');
      for (final model in _availableModels) {
        debugPrint(
            '  - ${model.name} (${model.category.displayName}) [${model.preferredFramework?.displayName ?? "Unknown"}] downloaded: ${model.isDownloaded}');
      }
    } catch (e) {
      debugPrint('❌ Failed to load models from SDK: $e');
      _errorMessage = 'Failed to load models: $e';
      _availableModels = [];
    }

    _currentModel = null;
    _isLoading = false;
    notifyListeners();
  }

  /// Convert SDK ModelInfo to app ModelInfo
  ModelInfo _convertSDKModel(sdk.ModelInfo sdkModel) {
    final framework = _convertFramework(sdkModel.framework);
    return ModelInfo(
      id: sdkModel.id,
      name: sdkModel.name,
      category: _convertCategory(sdkModel.category),
      format: _convertFormat(sdkModel.format),
      downloadURL: sdkModel.downloadURL?.toString(),
      localPath: sdkModel.localPath?.toFilePath(),
      memoryRequired: sdkModel.downloadSize,
      compatibleFrameworks: [framework],
      preferredFramework: framework,
      supportsThinking: sdkModel.supportsThinking,
    );
  }

  /// Convert SDK ModelCategory to app ModelCategory
  ModelCategory _convertCategory(sdk.ModelCategory sdkCategory) {
    switch (sdkCategory) {
      case sdk.ModelCategory.language:
        return ModelCategory.language;
      case sdk.ModelCategory.multimodal:
        return ModelCategory.multimodal;
      case sdk.ModelCategory.speechRecognition:
        return ModelCategory.speechRecognition;
      case sdk.ModelCategory.speechSynthesis:
        return ModelCategory.speechSynthesis;
      case sdk.ModelCategory.vision:
        return ModelCategory.vision;
      case sdk.ModelCategory.imageGeneration:
        return ModelCategory.imageGeneration;
      case sdk.ModelCategory.audio:
        return ModelCategory.audio;
      case sdk.ModelCategory.embedding:
        return ModelCategory.embedding;
    }
  }

  /// Convert SDK ModelFormat to app ModelFormat
  ModelFormat _convertFormat(sdk.ModelFormat sdkFormat) {
    switch (sdkFormat) {
      case sdk.ModelFormat.gguf:
        return ModelFormat.gguf;
      case sdk.ModelFormat.onnx:
      case sdk.ModelFormat.ort:
        return ModelFormat.onnx;
      case sdk.ModelFormat.bin:
        return ModelFormat.bin;
      case sdk.ModelFormat.unknown:
        return ModelFormat.unknown;
    }
  }

  /// Convert SDK InferenceFramework to app LLMFramework
  LLMFramework _convertFramework(sdk.InferenceFramework sdkFramework) {
    switch (sdkFramework) {
      case sdk.InferenceFramework.llamaCpp:
        return LLMFramework.llamaCpp;
      case sdk.InferenceFramework.foundationModels:
        return LLMFramework.foundationModels;
      case sdk.InferenceFramework.onnx:
        return LLMFramework.onnxRuntime;
      case sdk.InferenceFramework.systemTTS:
        return LLMFramework.systemTTS;
      default:
        return LLMFramework.unknown;
    }
  }

  /// Convert app LLMFramework to SDK InferenceFramework
  sdk.InferenceFramework _convertToSDKFramework(LLMFramework framework) {
    switch (framework) {
      case LLMFramework.llamaCpp:
        return sdk.InferenceFramework.llamaCpp;
      case LLMFramework.foundationModels:
        return sdk.InferenceFramework.foundationModels;
      case LLMFramework.onnxRuntime:
        return sdk.InferenceFramework.onnx;
      case LLMFramework.systemTTS:
        return sdk.InferenceFramework.systemTTS;
      case LLMFramework.mediaPipe:
      case LLMFramework.whisperKit:
      case LLMFramework.unknown:
        return sdk.InferenceFramework.unknown;
    }
  }

  /// Get available frameworks based on registered models
  Future<void> loadAvailableFrameworks() async {
    try {
      // Extract unique frameworks from available models
      final frameworks = <LLMFramework>{};
      for (final model in _availableModels) {
        if (model.preferredFramework != null) {
          frameworks.add(model.preferredFramework!);
        }
        frameworks.addAll(model.compatibleFrameworks);
      }
      _availableFrameworks = frameworks.toList();
      debugPrint(
          '✅ Available frameworks: ${_availableFrameworks.map((f) => f.displayName).join(", ")}');
      notifyListeners();
    } catch (e) {
      debugPrint('❌ Failed to load frameworks: $e');
      _availableFrameworks = [];
      notifyListeners();
    }
  }

  /// Alias for loadModelsFromRegistry
  Future<void> loadModels() async {
    await loadModelsFromRegistry();
    await loadAvailableFrameworks();
  }

  /// Set current model
  void setCurrentModel(ModelInfo? model) {
    _currentModel = model;
    notifyListeners();
  }

  /// Select and load a model
  Future<void> selectModel(ModelInfo model) async {
    try {
      await loadModel(model);
      setCurrentModel(model);
      debugPrint('✅ Model ${model.name} selected and loaded');
    } catch (e) {
      _errorMessage = 'Failed to load model: $e';
      notifyListeners();
    }
  }

  /// Download a model using SDK DownloadService
  /// This is the proper implementation using the SDK's download functionality
  Future<void> downloadModel(
    ModelInfo model,
    void Function(double) progressHandler,
  ) async {
    if (_downloadingModels.contains(model.id)) {
      debugPrint('⚠️ Model ${model.id} is already downloading');
      return;
    }

    _downloadingModels.add(model.id);
    _downloadProgress[model.id] = 0.0;
    notifyListeners();

    try {
      debugPrint('📥 Starting download for model: ${model.name}');

      // Use SDK's public download API
      await for (final progress in sdk.RunAnywhere.downloadModel(model.id)) {
        final progressValue = progress.totalBytes > 0
            ? progress.bytesDownloaded / progress.totalBytes
            : 0.0;

        _downloadProgress[model.id] = progressValue;
        progressHandler(progressValue);
        notifyListeners();

        // Check if completed or failed
        if (progress.state.isCompleted) {
          debugPrint('✅ Download completed for model: ${model.name}');
          break;
        } else if (progress.state.isFailed) {
          throw Exception('Download failed');
        }
      }

      // Update model with local path after download
      await loadModelsFromRegistry();

      debugPrint('✅ Model ${model.name} download complete');
    } catch (e) {
      debugPrint('❌ Failed to download model ${model.id}: $e');
      _errorMessage = 'Download failed: $e';
    } finally {
      _downloadingModels.remove(model.id);
      _downloadProgress.remove(model.id);
      notifyListeners();
    }
  }

  /// Delete a downloaded model using SDK
  Future<void> deleteModel(ModelInfo model) async {
    try {
      debugPrint('🗑️ Deleting model: ${model.name}');

      // Use SDK's public delete API (now only takes modelId)
      await sdk.RunAnywhere.deleteStoredModel(model.id);

      // Refresh models from registry
      await loadModelsFromRegistry();

      debugPrint('✅ Model ${model.name} deleted successfully');
    } catch (e) {
      debugPrint('❌ Failed to delete model: $e');
      _errorMessage = 'Failed to delete model: $e';
      notifyListeners();
    }
  }

  /// Load a model into memory using SDK
  Future<void> loadModel(ModelInfo model) async {
    _isLoading = true;
    notifyListeners();

    try {
      debugPrint('⏳ Loading model: ${model.name}');

      // Use appropriate SDK method based on model category
      switch (model.category) {
        case ModelCategory.language:
          await sdk.RunAnywhere.loadModel(model.id);
          break;
        case ModelCategory.speechRecognition:
          await sdk.RunAnywhere.loadSTTModel(model.id);
          break;
        case ModelCategory.speechSynthesis:
          await sdk.RunAnywhere.loadTTSVoice(model.id);
          break;
        default:
          // Default to LLM model loading
          await sdk.RunAnywhere.loadModel(model.id);
      }

      _currentModel = model;
      debugPrint('✅ Model ${model.name} loaded successfully');
    } catch (e) {
      debugPrint('❌ Failed to load model ${model.id}: $e');
      _errorMessage = 'Failed to load model: $e';
      rethrow;
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  /// Unload the current model
  Future<void> unloadCurrentModel() async {
    if (_currentModel == null) return;

    _isLoading = true;
    notifyListeners();

    try {
      await sdk.RunAnywhere.unloadModel();
      _currentModel = null;
      debugPrint('✅ Model unloaded successfully');
    } catch (e) {
      debugPrint('❌ Failed to unload model: $e');
      _errorMessage = 'Failed to unload model: $e';
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  /// Add a custom model from URL using SDK
  Future<void> addModelFromURL({
    required String name,
    required String url,
    required LLMFramework framework,
    int? estimatedSize,
    bool supportsThinking = false,
  }) async {
    try {
      debugPrint('➕ Adding model from URL: $name');

      // Use SDK's public registration API
      final modelInfo = sdk.RunAnywhere.registerModel(
        name: name,
        url: Uri.parse(url),
        framework: _convertToSDKFramework(framework),
        modality: sdk.ModelCategory.language,
        supportsThinking: supportsThinking,
      );

      debugPrint(
          '✅ Registered model with SDK: ${modelInfo.name} (${modelInfo.id})');

      // Refresh models from registry
      await loadModelsFromRegistry();

      debugPrint('✅ Model $name added successfully');
    } catch (e) {
      debugPrint('❌ Failed to add model from URL: $e');
      _errorMessage = 'Failed to add model: $e';
      notifyListeners();
    }
  }

  /// Add an imported model
  Future<void> addImportedModel(ModelInfo model) async {
    await loadModelsFromRegistry();
  }

  /// Get models for a specific framework
  List<ModelInfo> modelsForFramework(LLMFramework framework) {
    return _availableModels.where((model) {
      if (framework == LLMFramework.foundationModels) {
        return model.preferredFramework == LLMFramework.foundationModels;
      }
      return model.compatibleFrameworks.contains(framework);
    }).toList();
  }

  /// Get models for a specific context
  List<ModelInfo> modelsForContext(ModelSelectionContext context) {
    return _availableModels.where((model) {
      return context.relevantCategories.contains(model.category);
    }).toList();
  }

  /// Clear error message
  void clearError() {
    _errorMessage = null;
    notifyListeners();
  }
}

File: ios/runanywhere-sdks-latest/examples/ios/RunAnywhereAI/RunAnywhereAI/Features/Models/ModelListViewModel.swift
//
//  ModelListViewModel.swift
//  RunAnywhereAI
//
//  Simplified version that uses SDK registry directly
//

import Foundation
import SwiftUI
import RunAnywhere
import Combine

@MainActor
class ModelListViewModel: ObservableObject {
    static let shared = ModelListViewModel()

    @Published var availableModels: [ModelInfo] = []
    @Published var currentModel: ModelInfo?
    @Published var isLoading = false
    @Published var errorMessage: String?

    private var cancellables = Set<AnyCancellable>()

    // MARK: - Initialization

    init() {
        // Subscribe to SDK events for model lifecycle updates
        subscribeToModelEvents()

        Task {
            await loadModelsFromRegistry()
        }
    }

    /// Subscribe to SDK events for real-time model state updates
    private func subscribeToModelEvents() {
        // Subscribe to LLM events via EventBus
        RunAnywhere.events.events
            .receive(on: DispatchQueue.main)
            .sink { [weak self] event in
                guard let self = self else { return }
                self.handleSDKEvent(event)
            }
            .store(in: &cancellables)
    }

    /// Handle SDK events to update model state
    private func handleSDKEvent(_ event: any SDKEvent) {
        // Events now come from C++ via generic BridgedEvent
        guard event.category == .llm else { return }

        let modelId = event.properties["model_id"] ?? ""

        switch event.type {
        case "llm_model_load_completed":
            // Find the matching model and set as current
            if let matchingModel = availableModels.first(where: { $0.id == modelId }) {
                currentModel = matchingModel
                print("✅ ModelListViewModel: Model loaded: \(matchingModel.name)")
            }
        case "llm_model_unloaded":
            if currentModel?.id == modelId {
                currentModel = nil
                print("ℹ️ ModelListViewModel: Model unloaded: \(modelId)")
            }
        default:
            break
        }
    }

    // MARK: - Methods

    /// Load models from SDK registry (no more hard-coded models)
    func loadModelsFromRegistry() async {
        isLoading = true
        errorMessage = nil

        do {
            // Get all models from SDK registry
            // This now includes:
            // 1. Models from remote configuration (if available)
            // 2. Models from framework adapters
            // 3. Models from local storage
            // 4. User-added models
            let allModels = try await RunAnywhere.availableModels()

            // Filter based on iOS version if needed
            var filteredModels = allModels

            // Filter out Foundation Models for older iOS versions
            if #unavailable(iOS 26.0) {
                filteredModels = allModels.filter { $0.framework != .foundationModels }
                print("iOS < 26 - Foundation Models not available")
            }

            availableModels = filteredModels
            print("Loaded \(availableModels.count) models from registry")

            for model in availableModels {
                print("  - \(model.name) (\(model.framework.displayName))")
            }

            // Sync currentModel with SDK's current model state
            await syncCurrentModelWithSDK()
        } catch {
            print("Failed to load models from SDK: \(error)")
            errorMessage = "Failed to load models: \(error.localizedDescription)"
            availableModels = []
        }

        isLoading = false
    }

    /// Sync current model state with SDK
    private func syncCurrentModelWithSDK() async {
        if let currentModelId = await RunAnywhere.getCurrentModelId(),
           let matchingModel = availableModels.first(where: { $0.id == currentModelId }) {
            currentModel = matchingModel
            print("✅ ModelListViewModel: Restored currentModel from SDK: \(matchingModel.name)")
        }
    }

    func setCurrentModel(_ model: ModelInfo?) {
        currentModel = model
    }

    /// Alias for loadModelsFromRegistry to match view calls
    func loadModels() async {
        await loadModelsFromRegistry()
    }

    /// Select and load a model
    func selectModel(_ model: ModelInfo) async {
        do {
            try await loadModel(model)
            setCurrentModel(model)

            // Post notification that model was loaded successfully
            await MainActor.run {
                NotificationCenter.default.post(
                    name: Notification.Name("ModelLoaded"),
                    object: model
                )
            }
        } catch {
            errorMessage = "Failed to load model: \(error.localizedDescription)"
            // Don't set currentModel if loading failed
        }
    }

    func downloadModel(_ model: ModelInfo) async throws {
        // Use the SDK's public download API
        let progressStream = try await RunAnywhere.downloadModel(model.id)

        // Wait for completion
        for await progress in progressStream {
            print("Download progress: \(Int(progress.overallProgress * 100))%")
            if progress.stage == .completed {
                break
            }
        }

        // Reload models after download
        await loadModelsFromRegistry()
    }

    func deleteModel(_ model: ModelInfo) async throws {
        try await RunAnywhere.deleteStoredModel(model.id, framework: model.framework)
        // Reload models after deletion
        await loadModelsFromRegistry()
    }

    func loadModel(_ model: ModelInfo) async throws {
        try await RunAnywhere.loadModel(model.id)
        currentModel = model
    }

    /// Add a custom model from URL
    func addModelFromURL(name: String, url: URL, framework: InferenceFramework, estimatedSize: Int64?) async {
        // Use SDK's registerModel method
        RunAnywhere.registerModel(
            name: name,
            url: url,
            framework: framework,
            memoryRequirement: estimatedSize
        )

        // Reload models to include the new one
        await loadModelsFromRegistry()
    }

    /// Add an imported model to the list
    func addImportedModel(_ model: ModelInfo) async {
        // Just reload the models - the SDK registry will pick up the new model
        await loadModelsFromRegistry()
    }
}
